godot-core/src/builtin/script.rs

```diff
@@ -0,0 +1,818 @@
+    // choice. The validity of the instance info struct is ensured by code generation.
+    unsafe {
+        sys::interface_fn!(script_instance_create)(
+            instance_ptr,
+            data_ptr as sys::GDExtensionScriptInstanceDataPtr,
+        )
+    }
+}
+
+mod script_instance_info {
+    use std::any::type_name;
+    use std::cell::{BorrowError, Ref, RefMut};
+    use std::ffi::c_void;
+    use std::mem::ManuallyDrop;
+
+    use crate::builtin::meta::GodotType;
+    use crate::builtin::{GString, StringName, Variant};
+    use crate::engine::{Object, ScriptLanguage};
+    use crate::obj::Gd;
+    use crate::private::handle_panic;
+    use crate::sys::{self, GodotFfi};
+
+    use super::{ScriptInstance, ScriptInstanceData};
+
+    fn borrow_instance_mut<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> RefMut<'a, T> {
+        unsafe { (*instance).inner.borrow_mut() }
+    }
+
+    fn borrow_instance<'a, T: ScriptInstance>(instance: *mut ScriptInstanceData<T>) -> Ref<'a, T> {
+        unsafe { (*instance).inner.borrow() }
+    }
+
+    fn try_borrow_instance<'a, T: ScriptInstance>(
+        instance: *mut ScriptInstanceData<T>,
+    ) -> Result<Ref<'a, T>, BorrowError> {
+        unsafe { (*instance).inner.try_borrow() }
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a vallid instance pointer.
+    fn instance_data_as_script_instance<T: ScriptInstance>(
+        p_instance: sys::GDExtensionScriptInstanceDataPtr,
+    ) -> *mut ScriptInstanceData<T> {
+        p_instance as *mut ScriptInstanceData<T>
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid const string name pointer.
+    unsafe fn transfer_string_name_from_godot(
+        p_name: sys::GDExtensionConstStringNamePtr,
+    ) -> StringName {
+        StringName::from_arg_ptr(
+            p_name as sys::GDExtensionTypePtr,
+            sys::PtrcallType::Standard,
+        )
+    }
+
+    fn transfer_bool_to_godot(value: bool) -> u8 {
+        value as u8
+    }
+
+    /// # Safety
+    ///
+    /// - We expect the engine to provide a valid variant pointer the return value can be moved into.
+    unsafe fn transfer_variant_to_godot(variant: Variant, return_ptr: sys::GDExtensionVariantPtr) {
+        variant.move_return_ptr(
+            return_ptr as sys::GDExtensionTypePtr,
+            sys::PtrcallType::Standard,
+        )
+    }
+
+    /// # Safety
+    ///
+    /// - The engine has to provide a valid `*mut u32` pointer for the list length.
+    /// - The list pointer is guaranteed to point to a list that has an equal length and capacity.
+    unsafe fn transfer_ptr_list_to_godot<T>(ptr_list: Vec<T>, list_length: *mut u32) -> *const T {
+        *list_length = ptr_list.len() as u32;
+
+        let ptr = Box::into_raw(ptr_list.into_boxed_slice());
+
+        (*ptr).as_ptr()
+    }
```

Comment:
> Also you should put the unsafe operations in unsafe blocks with safety comments. It makes it easier to see where we're doing potentially UB operations. Rust currently accepts not doing it, but it will become a warning in 2024 rust. See https://rust-lang.github.io/rfcs/2585-unsafe-block-in-unsafe-fn.html.

Oh, this is interesting! I read this quite a while ago but wasn't aware it's now being integrated to Rust 2024.

Either way, I'm not sure if we should already follow this advice -- gdext is full of functions that don't do this, and moving to it would need bigger coordination, ideally tool-assisted. There seems to be [`unsafe_op_in_unsafe_fn`](https://github.com/rust-lang/rust/pull/79208) which could help. But we need to analyze the impact on ergonomics as well -- since we have a ton of very short functions, it might hurt more than it helps.

Reply:
> Either way, I'm not sure if we should already follow this advice -- gdext is full of functions that don't do this, and moving to it would need bigger coordination, ideally tool-assisted. There seems to be [`unsafe_op_in_unsafe_fn`](https://github.com/rust-lang/rust/pull/79208) which could help. But we need to analyze the impact on ergonomics as well -- since we have a ton of very short functions, it might hurt more than it helps.

That's fair, but at least in functions that are not one-liners i think we can pre-emptively do it. And i think it is useful to have the lint in general especially for bigger functions. which would mean we either need to allow the lint specifically in smaller functions (which is basically the same as adding the unsafe), or forbid the lint in bigger functions (which doesn't really help).

Reply:
Ah yeah, that makes sense. The pointer has to point to something valid, but the previous (dereferenced) value at that address doesn't matter.

Reply:
@lilizoey yeah, but let's not introduce new rules in this already huge and FFI-heavy PR. 

We already impose enough requirements just with _current_ rules, so I'd take it offline 🙂

Reply:
I tried to clarify the comments a bit.